File: src/WslTerminal/Ui/WslFiles.cs
using System.IO;

namespace WslTerminal.Ui;

/// <summary>
/// Bridges WSL (Linux) paths to Windows so the file sidebar can list and read
/// them without spawning a process: WSL exposes its filesystem to Windows at
/// <c>\\wsl.localhost\&lt;distro&gt;\…</c> (older builds: <c>\\wsl$\&lt;distro&gt;</c>),
/// so a Linux path like <c>/home/u/p</c> maps to <c>\\wsl.localhost\Ubuntu\home\u\p</c>.
/// </summary>
internal static class WslFiles
{
    /// <summary>Map a Linux path in <paramref name="distro"/> to its Windows UNC form.</summary>
    public static string ToUnc(string distro, string linuxPath)
    {
        string rel = linuxPath.Replace('/', '\\').TrimStart('\\');
        return $@"\\wsl.localhost\{distro}\{rel}";
    }

    /// <summary>One entry in a directory listing.</summary>
    public sealed record Entry(string Name, string LinuxPath, bool IsDir, long Size);

    /// <summary>
    /// List <paramref name="linuxDir"/>'s immediate children (dirs first, then files,
    /// each alphabetical, case-insensitive). Returns empty on any I/O error so the
    /// UI never throws. Dotfiles are included.
    /// </summary>
    public static List<Entry> List(string distro, string linuxDir)
    {
        var result = new List<Entry>();
        try
        {
            string unc = ToUnc(distro, linuxDir);
            var di = new DirectoryInfo(unc);
            if (!di.Exists) return result;

            var dirs = new List<Entry>();
            var files = new List<Entry>();
            string baseLinux = linuxDir.TrimEnd('/');
            foreach (var fsi in di.EnumerateFileSystemInfos())
            {
                bool isDir = (fsi.Attributes & FileAttributes.Directory) != 0;
                string childLinux = baseLinux + "/" + fsi.Name;
                long size = isDir ? 0 : SafeLen(fsi);
                var e = new Entry(fsi.Name, childLinux, isDir, size);
                (isDir ? dirs : files).Add(e);
            }
            dirs.Sort((a, b) => string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase));
            files.Sort((a, b) => string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase));
            result.AddRange(dirs);
            result.AddRange(files);
        }
        catch { /* permission / not-a-dir / disconnected: empty list */ }
        return result;
    }

    private static long SafeLen(FileSystemInfo fsi)
    {
        try { return fsi is FileInfo fi ? fi.Length : 0; }
        catch { return 0; }
    }

    /// <summary>Read a file's bytes (capped) for preview. Null on error.</summary>
    public static byte[]? ReadBytes(string distro, string linuxPath, int maxBytes = 2 * 1024 * 1024)
    {
        try
        {
            string unc = ToUnc(distro, linuxPath);
            using var fs = new FileStream(unc, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            int n = (int)Math.Min(fs.Length, maxBytes);
            var buf = new byte[n];
            int read = 0;
            while (read < n)
            {
                int r = fs.Read(buf, read, n - read);
                if (r <= 0) break;
                read += r;
            }
            return read == n ? buf : buf[..read];
        }
        catch { return null; }
    }

    /// <summary>The file's full size in bytes (-1 on error). Used to detect when a
    /// preview was truncated at the read cap, so we don't save a partial file.</summary>
    public static long FileLength(string distro, string linuxPath)
    {
        try { return new FileInfo(ToUnc(distro, linuxPath)).Length; }
        catch { return -1; }
    }

    /// <summary>Write UTF-8 text back to a WSL file (no BOM, overwriting). Returns
    /// false on any I/O error. The write goes through the \\wsl.localhost share, an
    /// in-place overwrite that keeps the existing inode/permissions.</summary>
    public static bool WriteText(string distro, string linuxPath, string text)
    {
        try
        {
            var bytes = new System.Text.UTF8Encoding(false).GetBytes(text);
            using var fs = new FileStream(ToUnc(distro, linuxPath), FileMode.Create, FileAccess.Write, FileShare.Read);
            fs.Write(bytes, 0, bytes.Length);
            return true;
        }
        catch { return false; }
    }

    /// <summary>True if the first chunk of bytes looks like binary (has a NUL).</summary>
    public static bool LooksBinary(ReadOnlySpan<byte> bytes)
    {
        int n = Math.Min(bytes.Length, 8000);
        for (int i = 0; i < n; i++) if (bytes[i] == 0) return true;
        return false;
    }

    private static readonly HashSet<string> ImageExt = new(StringComparer.OrdinalIgnoreCase)
        { ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".ico", ".webp" };

    public static bool IsImage(string name) => ImageExt.Contains(Path.GetExtension(name));

    public static string HumanSize(long bytes)
    {
        string[] u = { "B", "KB", "MB", "GB", "TB" };
        double s = bytes; int i = 0;
        while (s >= 1024 && i < u.Length - 1) { s /= 1024; i++; }
        return i == 0 ? $"{bytes} B" : $"{s:0.#} {u[i]}";
    }
}

File: rust/wslterm/src/wslfiles.rs
//! Bridge WSL (Linux) paths to Windows via the `\\wsl.localhost\<distro>\…`
//! share, so the sidebar/editor can list, read and write files without spawning
//! a process. Ports `src/WslTerminal/Ui/WslFiles.cs`.

use std::path::PathBuf;

/// Map a Linux path to its Windows UNC form under the distro share.
pub fn to_unc(distro: &str, linux_path: &str) -> PathBuf {
    let rel = linux_path.replace('/', "\\");
    let rel = rel.trim_start_matches('\\');
    PathBuf::from(format!(r"\\wsl.localhost\{distro}\{rel}"))
}

/// One directory entry.
pub struct Entry {
    pub name: String,
    pub linux_path: String,
    pub is_dir: bool,
}

/// List a Linux directory's immediate children (dirs first, then files, each
/// case-insensitively sorted). Empty on any I/O error. `show_hidden` includes
/// dotfiles.
pub fn list(distro: &str, linux_dir: &str, show_hidden: bool) -> Vec<Entry> {
    let mut dirs = Vec::new();
    let mut files = Vec::new();
    let base = linux_dir.trim_end_matches('/');
    let rd = match std::fs::read_dir(to_unc(distro, linux_dir)) {
        Ok(rd) => rd,
        Err(_) => return Vec::new(),
    };
    for ent in rd.flatten() {
        let name = ent.file_name().to_string_lossy().to_string();
        if !show_hidden && name.starts_with('.') {
            continue;
        }
        let is_dir = ent.file_type().map(|t| t.is_dir()).unwrap_or(false);
        let linux_path = format!("{base}/{name}");
        let e = Entry { name, linux_path, is_dir };
        if is_dir {
            dirs.push(e);
        } else {
            files.push(e);
        }
    }
    let key = |e: &Entry| e.name.to_lowercase();
    dirs.sort_by_key(key);
    files.sort_by_key(key);
    dirs.append(&mut files);
    dirs
}

/// Read a file's bytes, capped at `max` (default 2 MiB). `None` on error.
pub fn read_bytes(distro: &str, linux_path: &str, max: usize) -> Option<Vec<u8>> {
    let bytes = std::fs::read(to_unc(distro, linux_path)).ok()?;
    if bytes.len() > max {
        Some(bytes[..max].to_vec())
    } else {
        Some(bytes)
    }
}

/// Overwrite a WSL file with UTF-8 text (in place; keeps inode/permissions).
pub fn write_text(distro: &str, linux_path: &str, text: &str) -> bool {
    std::fs::write(to_unc(distro, linux_path), text.as_bytes()).is_ok()
}

/// True if the first chunk looks binary (contains a NUL).
pub fn looks_binary(bytes: &[u8]) -> bool {
    bytes.iter().take(8000).any(|&b| b == 0)
}
